src/sage/graphs/graph.py

```diff
@@ -4328,6 +4328,228 @@ def weight(x):
+            sage: S == set(G.vertices())
+            True
+
+        For a matching covered graph `G` with a perfect matching `M` and for
+        some vertex `u` with `v` being its `M`-matched neighbor, each neighbor
+        of `v` is reachable from `u` through an even length `M`-alternating
+        path as described above::
+
+            sage: # K(4) ☉ K(3, 3) is matching covered
+            sage: G = Graph()
+            sage: G.add_edges([
+            ....:    (0, 2), (0, 3), (0, 4), (1, 2),
+            ....:    (1, 3), (1, 4), (2, 5), (3, 6),
+            ....:    (4, 7), (5, 6), (5, 7), (6, 7)
+            ....: ])
+            sage: M = Graph(G.matching())
+            sage: G.is_matching_covered(M)
+            True
+            sage: u = 0
+            sage: v = next(M.neighbor_iterator(u))
+            sage: S = G.M_alternating_even_mark(u, M)
+            sage: (set(G.neighbor_iterator(v))).issubset(S)
+            True
+
+        For a bicritical graph `G` with a perfect matching `M` and for some
+        vertex `u` with its `M`-matched neighbor being `v`, each vertex of the
+        graph distinct from `v` is reachable from `u` through an even length
+        `M`-alternating path as described above::
+
+            sage: # Petersen graph is bicritical
+            sage: G = graphs.PetersenGraph()
+            sage: M = Graph(G.matching())
+            sage: G.is_bicritical(M)
+            True
+            sage: u = random.choice(G.vertices())  # needs random
+            sage: v = next(M.neighbor_iterator(u))
+            sage: S = G.M_alternating_even_mark(u, M)
+            sage: S == (set(G.vertices()) - {v})
+            True
+
+        TESTS:
+
+        Giving a wrong vertex::
+
+            sage: G = graphs.HexahedralGraph()
+            sage: M = G.matching()
+            sage: u = G.order()
+            sage: S = G.M_alternating_even_mark(u, M)
+            Traceback (most recent call last):
+            ...
+            ValueError: '8' is not a vertex of the graph
+
+        Giving a wrong matching::
+
+            sage: G = graphs.CompleteGraph(6)
+            sage: M = [(0, 1), (0, 2)]
+            sage: u = 0
+            sage: S = G.M_alternating_even_mark(u, M)
+            Traceback (most recent call last):
+            ...
+            ValueError: the input is not a matching
+            sage: G = graphs.CompleteBipartiteGraph(3, 3)
+            sage: M = [(2*i, 2*i + 1) for i in range(4)]
+            sage: u = 0
+            sage: S = G.M_alternating_even_mark(u, M)
+            Traceback (most recent call last):
+            ...
+            ValueError: the input is not a matching of the graph
+
+        REFERENCES:
+
+        - [LR2004]_
+
+        .. SEEALSO::
+            :meth:`~sage.graphs.graph.Graph.is_factor_critical`,
+            :meth:`~sage.graphs.graph.Graph.is_matching_covered`,
+            :meth:`~sage.graphs.graph.Graph.is_bicritical`
+        """
+
+        G = self.to_simple()
```

Comment:
This can be done after some tests

Reply:
If we allow multiple edges in _G_, then the following error happens:

Consider the following example:

```
sage: G = graphs.CompleteBipartiteGraph(3, 3)
sage: G.allow_multiple_edges(True)
sage: G.add_edge(0, 3)
sage: M = G.matching()
sage: u = 0
sage: S = G.M_alternating_even_mark(u, M)
sage: S
```

It shows the following `ValueError` for `G.M_alternating_even_mark(u, M)`:

```
{
	"name": "ValueError",
	"message": "This method is not known to work on graphs with multiedges. Perhaps this method can be updated to handle them, but in the meantime if you want to use it please disallow multiedges using allow_multiple_edges().",
	"stack": "---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/tmp/ipykernel_987822/3075726170.py in ?()
      1 u = Integer(0)
----> 2 G.M_alternating_even_mark(u, M)

~/open-source/sage/src/sage/graphs/graph.py in ?(self, vertex, matching)
   4480         # The input matching must be a valid matching of the graph
   4481         M = Graph(matching)
   4482         if any(d != 1 for d in M.degree()):
   4483             raise ValueError(\"the input is not a matching\")
-> 4484         if not M.is_subgraph(self, induced=False):
   4485             raise ValueError(\"the input is not a matching of the graph\")
   4486 
   4487         # Build an M-alternating tree T rooted at vertex

~/open-source/sage/src/sage/graphs/generic_graph.py in ?(self, other, induced, up_to_isomorphism)
  15813         if not all(other.has_vertex(v) for v in self.vertex_iterator()):
  15814             return False
  15815 
  15816         self._scream_if_not_simple()
> 15817         other._scream_if_not_simple()
  15818 
  15819         if induced:
  15820             # Check whether ``self`` is contained in ``other``

~/open-source/sage/src/sage/graphs/generic_graph.py in ?(self, allow_loops, allow_multiple_edges)
   1495             msg = (\"This method is not known to work on graphs with \" + name + \". \"
   1496                    \"Perhaps this method can be updated to handle them, but in the \" +
   1497                    \"meantime if you want to use it please disallow \" + name + \" using \" +
   1498                    functions + \".\")
-> 1499             raise ValueError(msg)

ValueError: This method is not known to work on graphs with multiedges. Perhaps this method can be updated to handle them, but in the meantime if you want to use it please disallow multiedges using allow_multiple_edges()."

}
```

Reply:
You can force the input graph to be simple using `G._scream_if_not_simple()`. This way the method will be valid only for simple graphs.